Feature/lp 21 community

AndresIvanMedinaH opened the pull request:
**Task’s name**
Community - UI
      **Task’s link to navigate to jira**
This UI must show a list of elements that will lead the user to a different kind of ammenities

      **General task’s description of what was added, deleted or updated in the code.**
    The view must be implemented according to the design provided by the design team.

    Each card must contain the name of the section and an icon.

    The list of community sections must be displayed in a grid format with two cards per row.

    The view must be responsive and adapt to different screen sizes and orientations.

    The font and colors used should be consistent with the design of the rest of the application.


      **Attachment file of how looks your code in the app or in the preview if it is necessary.** 
![communityUI](https://user-images.githubusercontent.com/128720218/233200776-e2057aca-25ce-4f5c-a658-2031e43345d9.jpg)

AndresIvanMedinaH committed:
Merge branch 'Develop' into LP-160-BoldWeight-LpIconTextCard

AndresIvanMedinaH committed:
feature/LP-21-Community

Added FontWeigth property to component
-Removed files from Context
-Merged from develop to retrieve updates

AndresIvanMedinaH committed:
feature/LP-21-Community
Changed parameters and added new ones
Enabled = false
Keyboard to accept alphanumeric
Changed name to some more generic
update to file Card due to dependencies

AndresIvanMedinaH committed:
Merge branch 'Develop' into feature/LP-21-Community

AndresIvanMedinaH committed:
feature/LP-21-Community
Changed parameters and added new ones
Enabled = false
Keyboard to accept alphanumeric
Changed name to some more generic
update to file Card due to dependencies
update graddle libraries for feature

AndresIvanMedinaH committed:
feature/LP-21-Community
Changed parameters and added new ones
Enabled = false
Keyboard to accept alphanumeric
Changed name to some more generic
update to file Card due to dependencies
update graddle libraries for feature

AndresIvanMedinaH committed:
feature/LP-21-Community
Changed parameters and added new ones
Enabled = false
Keyboard to accept alphanumeric
Changed name to some more generic
update to file Card due to dependencies
update graddle libraries for feature

AndresIvanMedinaH committed:
Merge remote-tracking branch 'Irving/feature/LP-21-Community' into feature/LP-21-Community

AndresIvanMedinaH committed:
Update build.graddle file to fix Compose navigation issue

dmroldanch commented on feature/community/build.gradle:
```diff
@@ -28,8 +30,24 @@ android {
 
 dependencies {
+    implementation libs.bundles.compose
+    implementation libs.compose.navigation
```
You don't have to import this library like this, because you already have the whole compose bundle, where you already have it

AndresIvanMedinaH replied:
yes It was as a test, I already removed from, I understand compose and navigation is already bundled inside the lib.bundles.compose variable. thanks

dmroldanch commented on feature/community/build.gradle:
```diff
@@ -28,8 +30,24 @@ android {
 
+//Modules
+    implementation project(":core:designsystem")
+    implementation project(":app")
```
You shouldn't import the app module

AndresIvanMedinaH replied:
this one I know, removed

dmroldanch commented on feature/community/build.gradle:
```diff
@@ -28,8 +30,24 @@ android {
+    implementation libs.bundles.compose
+    implementation libs.compose.navigation
+    implementation libs.core
+    implementation libs.lifecycle
```
what are you using lifecycle for?

AndresIvanMedinaH replied:
removed

dmroldanch commented on feature/community/build.gradle:
```diff
@@ -28,8 +30,21 @@ android {
 
 dependencies {
+    implementation libs.bundles.compose
+    implementation libs.core
```
what are you using core for?

AndresIvanMedinaH replied:
removed

dmroldanch commented on feature/community/build.gradle:
```diff
@@ -1,6 +1,8 @@
-    id 'com.android.library'
+    id 'com.android.application'
     id 'org.jetbrains.kotlin.android'
+    id 'kotlin-kapt'
```
You have to remove everything related to Hilt, we haven't started with the injection dependencies yet.

AndresIvanMedinaH replied:
removed

dmroldanch commented on feature/community/build.gradle:
```diff
@@ -28,8 +30,21 @@ android {
+    buildFeatures {
+        compose true
+    }
+    composeOptions {
```
Why are you using this?  

`kotlinCompilerExtensionVersion '1.2.0'`

AndresIvanMedinaH committed:
Update build.graddle file to fix Compose navigation issue

AndresIvanMedinaH committed:
Update build.graddle file to fix Compose navigation issue

AndresIvanMedinaH commented:
this is fhe first version of gradle file that I have used to push ... 
uploading again 

plugins {
    id 'com.android.application'
    id 'org.jetbrains.kotlin.android'
}

android {
    namespace 'com.iteneum.community'
    compileSdk 33

    defaultConfig {
        minSdk 25
        targetSdk 33

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
        consumerProguardFiles "consumer-rules.pro"
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = '1.8'
    }
    buildFeatures {
        compose true
    }

}

dependencies {
    implementation libs.bundles.compose


//Modules
    implementation project(":core:designsystem")
}

AndresIvanMedinaH commented:
this is app graddle.build: 

Should I remove all from here too?

plugins {
    id 'com.android.application'
    id 'org.jetbrains.kotlin.android'
    id 'kotlin-kapt'
    id 'com.google.dagger.hilt.android'
}

android {
    namespace 'com.iteneum.leasepert'
    compileSdk 33

    defaultConfig {
        applicationId "com.iteneum.leasepert"
        minSdk 28
        targetSdk 33
        versionCode 1
        versionName "1.0"

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
        vectorDrawables {
            useSupportLibrary true
        }
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = '1.8'
    }
    buildFeatures {
        compose true
    }
    composeOptions {
        kotlinCompilerExtensionVersion '1.2.0'
    }
    packagingOptions {
        resources {
            excludes += '/META-INF/{AL2.0,LGPL2.1}'
        }
    }
}

dependencies {

    implementation libs.bundles.compose
    implementation libs.core
    implementation libs.lifecycle
    implementation libs.hiltAndroid
    kapt libs.hiltCompiler

    implementation project(":core:designsystem")
    implementation project(":feature:example")
    implementation project(":feature:office")
}

irv205 commented on feature/community/src/main/java/com/iteneum/community/domain/CardMenuItem.kt:
- Naming convention: The enum values are in all caps, which is the recommended naming convention for constants in Kotlin. However, the naming could be improved by separating the words with underscores, like "COMMUNITY_WALL" instead of "COMMUNITYWALL".

- Ordering: The enum values are listed in no particular order. It's a good practice to order them alphabetically to make them easier to read and maintain.

- Consider adding a description or comment for the enum class to clarify its purpose:

- Consider renaming the CardMenuItem class to a more descriptive name that reflects the fact that it represents a list of community sections. For example, CommunitySectionList

irv205 commented on feature/community/src/main/java/com/iteneum/community/domain/CardModel.kt:
Naming conventions: The name CardModel is good, but you can improve it by making it more descriptive. For example, you could name it CommunitySectionModel to indicate that it represents a section of the community.

AndresIvanMedinaH replied:
renaming to CommunitySectionModel, even this is not a section but a Card

irv205 replied:
It's actually a model, a card is a component

AndresIvanMedinaH replied:
it is a model  cos of the data class feature but, represents the info in the Card, so what about to name it as something as
CommunityCardsModel

AndresIvanMedinaH replied:
it is a model  cos of the data class feature but, represents the info in the Card, so what about to name it as something as
CommunityCardsModel or CommunityCarDataModel

AndresIvanMedinaH replied:
Fixed and waiting for push

irv205 commented on feature/community/src/main/java/com/iteneum/community/domain/CardMenuItem.kt:
```diff
@@ -0,0 +1,9 @@
+package com.iteneum.community.domain
+
+enum class CardMenuItem {
```
- Naming convention: The enum values are in all caps, which is the recommended naming convention for constants in Kotlin. However, the naming could be improved by separating the words with underscores, like "COMMUNITY_WALL" instead of "COMMUNITYWALL".

- Ordering: The enum values are listed in no particular order. It's a good practice to order them alphabetically to make them easier to read and maintain.

- Consider adding a description or comment for the enum class to clarify its purpose:

- Consider renaming the CardMenuItem class to a more descriptive name that reflects the fact that it represents a list of community sections. For example, CommunitySectionList

AndresIvanMedinaH replied:
will not change this, cos of sealed class that will be used

irv205 replied:
Being something that is kept in the project, we must try to keep the code clean, always following the best practices, update the observations made in that class.

AndresIvanMedinaH replied:
Sorry I was not clear enough  i am not using enum class anymore due to sealed class, the other suggestions will read it carefully and take care of the things that have to change,  please consider open different comments for different changes  this one is cluttered of information  did not read the other bullets

irv205 replied:
Update your branch with the new changes, otherwise for me it's as if you haven't made any changes yet

AndresIvanMedinaH replied:
Ordering: It is ordered as it is needed to be presented in the presentation, even it is not necessary and filtered through a when  component, but will be ordered in ascendent  from A to Z

adding  description

Renamed to  CommunitySectionList

irv205 commented on feature/community/src/main/java/com/iteneum/community/presentation/CommunityComposable.kt:
```diff
@@ -0,0 +1,72 @@
+ * @author Andrés Ivan Medina Herrera
+ */
+@Composable
+fun CommunitylistUI() {
```
- The function name should follow the camelCase convention and start with a lowercase letter, for example, communityListView.

AndresIvanMedinaH replied:
it is a composable function, are you sure, it was suggested in here but renaming

AndresIvanMedinaH replied:
please review your sources, as it is suggested in pascalCase it has to start with capital letter https://www.valueof.io/blog/compose-composable-return-types

irv205 replied:
No, it is not a good practice to start the name of a function with a capital letter in Kotlin. According to Kotlin's coding conventions, function names should start with a lowercase letter and follow camelCase notation. This convention makes it easier to differentiate functions from classes and other types in the code.

For example, instead of naming a function "MyFunction", it should be named "myFunction".

![image](https://user-images.githubusercontent.com/42255248/233671155-4bb00c61-6cc1-4ac9-aaee-a681a13460d5.png)

Try to verify your information with the official documentation:
https://kotlinlang.org/docs/coding-conventions.html#function-names

AndresIvanMedinaH replied:
those are not composable functions those are regular functions

irv205 replied:
my mistake, you are right

irv205 commented on feature/community/src/main/java/com/iteneum/community/presentation/CommunityComposable.kt:
```diff
@@ -0,0 +1,72 @@
+import com.iteneum.designsystem.theme.LeasePertTheme
+
+/**
+ * This function creates a view that displays a list of community sections.
```
The function should have a comment explaining its purpose, but the comment in the current code is too verbose and does not add much value. A better comment could be something like: "This function creates a Composable UI that displays a list of community sections."

AndresIvanMedinaH replied:
changing "a view"   for a less verbose phrase  " a composable UI"

irv205 replied:
In compose all the UI is handled as View, if you have doubts about that, please look into compose.

AndresIvanMedinaH replied:
but basically this is a function not a view by itself, the activity or fragment in chage of the view is itself,    cos if the view has different composables included then that logic basically fell appart

irv205 commented on feature/community/src/main/java/com/iteneum/community/presentation/CommunityComposable.kt:
```diff
@@ -0,0 +1,72 @@
+ */
+@Composable
+fun CommunitylistUI() {
+    val cardWith = LeasePertTheme.sizes.extraSize124
```
The variable names are mostly clear and follow the camelCase convention, except for **cardWith**, which should be renamed to **cardWidth**.

AndresIvanMedinaH replied:
changing the typo
cardWith      to   cardWidth

carlosph816 commented on feature/community/src/main/java/com/iteneum/community/presentation/CommunityComposable.kt:
```diff
@@ -0,0 +1,72 @@
+            .padding(top = cardPadding20, start = cardPadding10, end = cardPadding10)
+    ) {
+        Text(
+            text = "Community",
```
I think this text we can add it to the string resource

irv205 replied:
Must be added to string file @AndresIvanMedinaH

AndresIvanMedinaH replied:
What is the standard used in the other screens?   as this is a Type of thext that exists in other screens there should be a standard is not it?  @irv205

carlosph816 commented on feature/community/src/main/java/com/iteneum/community/presentation/CommunityComposable.kt:
```diff
@@ -0,0 +1,72 @@
+        LazyVerticalGrid(
+            columns = GridCells.Fixed(2), contentPadding = PaddingValues(cardPadding8)
+        ) {
+            items(communityCardButtons) { singleCommunityCardButton ->
```
As a suggestion, you could avoid long names "singleCommunityCardButton" for these cases, you can replace it for example "item" or "itemCard"

AndresIvanMedinaH replied:
this was done according to the previous suggestion to be more specific by @melissaquezada94

AndresIvanMedinaH replied:
please be more clear on the naming variables or stablish a guideline @irv205  this is the case I was talking about, when I asked for guidelines, the previous name of that variable was cardButtons

irv205 replied:
In the case that @carlosph816 mentions, it is correct, since we refer to an internal element in a list, in the case of functions and classes it is different.

AndresIvanMedinaH replied:
changing  "singleCommunityCardButton"   to a less descriptive "itemCard"

melissaquezada94 commented on feature/community/src/main/java/com/iteneum/community/presentation/CommunityComposable.kt:
```diff
@@ -0,0 +1,72 @@
+                LpIconTextCard(modifier = Modifier.size(width = cardWith, height = cardHeight),
+                    icon = singleCommunityCardButton.icon,
+                    description = stringResource(
+                        id = when (singleCommunityCardButton.cardName) {
```
Please use a Sealed class with Objects because would be easier if we decide to change the way of get the data

AndresIvanMedinaH replied:
Sealed class isnt part of actually more logic side than UI  please elaborate why to use it in this state, 'cos that is the plan for later

AndresIvanMedinaH replied:
@irv205  please confirm the use of the sealed class

melissaquezada94 replied:
> Sealed class isnt part of actually more logic side than UI please elaborate why to use it in this state, 'cos that is the plan for later

Because you can learn how to implement the sealed class and see the pros of it if we can get the data from the server

irv205 replied:
It would be better to implement it as @melissaquezada94 advises you, then do it, please.

AndresIvanMedinaH replied:
adding sealed class

 the doubts presented on whatsapp communication were not answered bud dodged, please elaborate on those comments to proceed on implementation  in a clear way 
 
 also as this is a dependency will need you  @irv205  to create the task with the clear description and acceptance criteria to proceed on the implementation.

irv205 replied:
After analyzing using when in this code is a good approach. The when expression is a concise and readable way of handling different cases based on the input value. In this case, we have a finite number of possible values for singleCommunityCardButton.cardName, so it's a good use case for when.

Using a sealed class could also work in this scenario, but it would add unnecessary complexity to the code, since we only need to handle a small set of cases. A sealed class is generally used when we have a larger set of possible values and need to perform different operations depending on the value type.

melissaquezada94 replied:
Well a sealed class can be use with a small set of cases as well an can reduce the `when` complexity and also other code complexities. 
I still think we can simplify the code and provide better flexibility in the future even if it takes more time to implement or do a research about how to use it. 
In general we want to use sealed classes instead of Enums. But if you consider to much complexity then we can leave it as enums.  @irv205

melissaquezada94 commented on feature/community/src/main/java/com/iteneum/community/presentation/CommunityComposable.kt:
```diff
@@ -0,0 +1,72 @@
+                            CardMenuItem.SERVICES -> R.string.communityui_cardbutton_services
+                        }
+                    ),
+                    onCardClick = { })
```
Please add a //TODO comment describing what is going to do this `onCardClick`

AndresIvanMedinaH replied:
Fixing

melissaquezada94 commented on feature/community/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,9 @@
+    <string name="communityui_cardbutton_ammenities">Amenities</string>
+    <string name="communityui_cardbutton_events">Events</string>
+    <string name="communityui_cardbutton_communityWall">Community Wall</string>
+    <string name="communityui_cardbutton_doItYourself">Do It Yourself</string>
```
Please follow the camel case structure: ``communityUi_cardButton_doItYourself

AndresIvanMedinaH replied:
Fixing

AndresIvanMedinaH committed:
Merge remote-tracking branch 'Irving/feature/LP-21-Community' into feature/LP-21-Community

# Conflicts:
#	feature/community/build.gradle

AndresIvanMedinaH committed:
LP-21-CommunityUI
# Conflicts:
#Fixed TODO comment
#Fixed Strings in camel case

AndresIvanMedinaH committed:
LP-21-CommunityUI
# Conflicts:
Renamed model to CommunityCardDataModel

Ordering: It is ordered as it is needed to be presented in the presentation, even it is not necessary and filtered through a when component, but will be ordered in ascendent from A to Z

adding description

Renamed to CommunitySectionList

changing part of the documentation comment from: "a view" To: " a composable UI"

changed typo from: cardWith to: cardWidth

added resource to  strings file <string name="uI_title_community">Community</string>

changing "singleCommunityCardButton" to a less descriptive "itemCard"

AndresIvanMedinaH committed:
LP-21-CommunityUI
# Conflicts:
Renamed model to CommunityCardDataModel

Ordering: It is ordered as it is needed to be presented in the presentation, even it is not necessary and filtered through a when component, but will be ordered in ascendent from A to Z

adding description

Renamed to CommunitySectionList

changing part of the documentation comment from: "a view" To: " a composable UI"

changed typo from: cardWith to: cardWidth

added resource to  strings file <string name="uI_title_community">Community</string>

changing "singleCommunityCardButton" to a less descriptive "itemCard"

AndresIvanMedinaH committed:
LP-21-CommunityUI
# Conflicts:
Renamed model to CommunityCardDataModel

Ordering: It is ordered as it is needed to be presented in the presentation, even it is not necessary and filtered through a when component, but will be ordered in ascendent from A to Z

adding description

Renamed to CommunitySectionList
changing part of the documentation comment from: "a view" To: " a composable UI"
changed typo from: cardWith to: cardWidth
added resource to  strings file <string name="uI_title_community">Community</string>
Renamed :  <string name="title_community">Community</string>
changing "singleCommunityCardButton" to a less descriptive "itemCard"

melissaquezada94 commented on feature/community/src/main/java/com/iteneum/community/domain/CommunityCardDataModel.kt:
```diff
@@ -0,0 +1,13 @@
+ *
+ * @author Andrés Ivan Medina Herrera
+ */
+data class CommunityCardDataModel(val icon: ImageVector, val cardName: CommunitySectionList)
```
If you use a sealed class instead an Enum, you don't need to implement all this stuff and your code would be readable and easier to understand also you'll be able to avoid boilerplate and it would look more elegant.

AndresIvanMedinaH replied:
@irv205  will it be or will not be implemented  get cleared to proceed on the fix please

melissaquezada94 replied:
I still think you don't need acceptance criteria to create a sealed class, I mean, did you needed acceptance criteria to create a enum class? I just wanted to implement the sealed class because is one of the features of kotlin and would be easier and less rewrite when we'll decide to get the data from the server. Also because is important to you learn it and talk in the interviews with more confidence.

melissaquezada94 commented on feature/community/src/main/java/com/iteneum/community/domain/CommunitySectionList.kt:
```diff
@@ -0,0 +1,11 @@
+ *
+ * @author Andrés Ivan Medina Herrera
+ */
+enum class CommunitySectionList {
```
Something like this would be a similar approach but with sealed classes and there's no need for acceptance criteria, it is just a different programming style using kotlin features.


![Screenshot 2023-04-20 at 7 03 50 PM](https://user-images.githubusercontent.com/105077839/233746353-ec1cf20e-3f14-49cc-839d-c134fa4be4f5.png)

![Screenshot 2023-04-20 at 7 03 35 PM](https://user-images.githubusercontent.com/105077839/233746358-8a1151b7-4408-41b8-9c27-f668e1df030c.png)

AndresIvanMedinaH replied:
the acceptance criteria is because a different component as a sealed class will imply new component decisions and connections  over the objects that will be handled as options for the sealed class, this decisions concern to the Project master and the architect so as much as I know @irv205  or @dmroldanch  should take the decision  but on previous talk  

they mentioned this so : check with them @melissaquezada94 
![imagen](https://user-images.githubusercontent.com/128720218/234021824-d8a6b38d-f370-4130-9da0-d68637148962.png)

melissaquezada94 commented on feature/community/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,10 @@
+    <string name="communityUi_cardButton_doItYourself">Do It Yourself</string>
+    <string name="communityUi_cardButton_services">Services</string>
+    <string name="communityUi_title_community">Community</string>
+    <string name="uI_title_community">Community</string>
```
You can name it just like `title_community`

AndresIvanMedinaH replied:
fixing according to the suggestion

AndresIvanMedinaH replied:
updated

AndresIvanMedinaH committed:
Merge remote-tracking branch 'Irving/feature/LP-21-Community' into feature/LP-21-Community

AndresIvanMedinaH committed:
LP-21-CommunityUI
# Conflicts:
Renamed model to CommunityCardDataModel

Ordering: It is ordered as it is needed to be presented in the presentation, even it is not necessary and filtered through a when component, but will be ordered in ascendent from A to Z

adding description

Renamed to CommunitySectionList
changing part of the documentation comment from: "a view" To: " a composable UI"
changed typo from: cardWith to: cardWidth
added resource to  strings file <string name="uI_title_community">Community</string>
Renamed :  <string name="title_community">Community</string>
changing "singleCommunityCardButton" to a less descriptive "itemCard"
